httpcore/_async/http_proxy.py

```diff
@@ -140,47 +133,52 @@ def __init__(
         connection = await self._get_connection_from_pool(origin)
 
         if connection is None:
-            connection = AsyncHTTPConnection(
-                origin=origin, http2=False, ssl_context=self._ssl_context,
+            # First, create a connection to the proxy server
+            proxy_connection = AsyncHTTPConnection(
+                origin=self.proxy_origin, http2=False, ssl_context=self._ssl_context,
             )
             async with self._thread_lock:
                 self._connections.setdefault(origin, set())
-                self._connections[origin].add(connection)
+                self._connections[origin].add(proxy_connection)
```

Comment:
I thought about that too, but it seemed to me the proxy connection _is_ the connection to the origin for the time it takes to establish the tunnel. In other words, if another coroutine attempts a request it should not attempt a second proxy connection.

I'm not 100% sure that's how the pooling works, though. What do you think?

Reply:
Yes, the connection pooling looks into `self._connections` for the same origin, and there are a few more conditions to reusing a connection now.

Specifically, if a connection is `ACTIVE` and non-HTTP/1.1 (as is probably the case for the duration of us setting up the tunnel) then we will create a new connection anyway.

This means a concurrent coroutine would try and set up a new proxy connection _anyway_ (because it is `ACTIVE` inside this section), instead of reusing the existing one (which we don't even want: AFAIU we should use the final `connection` for regular requests, not the `proxy_connection`).

Reply:
Ah right, I forgot about that. 👍